File: js/src/Vista.SDK/UniversalId.Builder.ts
import { ImoNumber, LocalIdBuilder } from '.';
import { LocalIdErrorBuilder } from './internal/LocalIdErrorBuilder';
import { ParsingState } from './types/LocalId';
import { VisVersion } from './VisVersion';

export class UniversalIdBuilder {
  public readonly namingEntity = 'data.dnv.com';
  private _localId?: LocalIdBuilder;
  public imoNumber?: ImoNumber;

  public get localId() {
    return this._localId;
  }

  public get isValid(): boolean {
    return !!this.imoNumber?.isValid && !!this.localId?.isValid;
  }

  public static create(visVersion: VisVersion) {
    return new UniversalIdBuilder().withLocalId(LocalIdBuilder.create(visVersion));
  }

  public equals(other?: UniversalIdBuilder): boolean {
    if (!other) return false;
    return this.imoNumber === other.imoNumber && !!this.localId?.equals(other.localId);
  }

  public clone() {
    return Object.assign(new UniversalIdBuilder(), this);
  }

  public validate(errorBuilder = new LocalIdErrorBuilder()): LocalIdErrorBuilder {
    if (!this.imoNumber?.isValid) {
      errorBuilder.push({ type: ParsingState.EmptyState, message: 'Missing or invalid IMO number' });
    }
    this._localId?.validate(errorBuilder);
    return errorBuilder;
  }

  public toString() {
    const builder: string[] = [];

    if (!this.imoNumber) throw new Error('Invalid Universal Id state: Missing IMO Number');
    if (!this.localId) throw new Error('Invalid Universal Id state: Missing LocalId');

    builder.push(this.namingEntity);
    builder.push('/');
    builder.push('IMO');
    builder.push(this.imoNumber.toString());
    builder.push('/');
    this.localId.toString(builder);

    return builder.join('');
  }

  public withLocalId(localId?: LocalIdBuilder) {
    return this.with(s => (s._localId = localId));
  }

  public withImoNumber(imoNumber?: ImoNumber) {
    return this.with(s => (s.imoNumber = imoNumber));
  }

  public with(u: (state: UniversalIdBuilder) => void): UniversalIdBuilder {
    const n = this.clone();
    u && u(n);
    return n;
  }
}

File: csharp/src/Vista.SDK/UniversalIdBuilder.cs
using Vista.SDK.Internal;

namespace Vista.SDK;

public sealed partial record class UniversalIdBuilder : IUniversalIdBuilder
{
    public static readonly string NamingEntity = "data.dnv.com";
    private LocalIdBuilder? _localId;
    public ImoNumber? ImoNumber { get; private init; }

    public LocalIdBuilder? LocalId => _localId;

    public bool IsValid => ImoNumber is not null && _localId is not null && _localId.IsValid;

    public static UniversalIdBuilder Create(VisVersion version) =>
        new UniversalIdBuilder().WithLocalId(LocalIdBuilder.Create(version));

    public bool Equals(UniversalIdBuilder? other)
    {
        if (other is null)
            return false;
        return ImoNumber == other.ImoNumber && LocalId == other.LocalId;
    }

    public UniversalId Build()
    {
        return new UniversalId(this);
    }

    public UniversalIdBuilder WithLocalId(in LocalIdBuilder localId)
    {
        var universalIdbuilder = TryWithLocalId(localId, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithLocalId));

        return universalIdbuilder;
    }

    public UniversalIdBuilder WithoutLocalId() => this with { _localId = null };

    public UniversalIdBuilder TryWithLocalId(in LocalIdBuilder? localId, out bool succeeded)
    {
        if (localId == null)
        {
            succeeded = false;
            return this;
        }
        succeeded = true;
        return this with { _localId = localId };
    }

    public UniversalIdBuilder TryWithLocalId(in LocalIdBuilder? localId) => TryWithLocalId(localId, out _);

    public UniversalIdBuilder WithImoNumber(in ImoNumber imoNumber)
    {
        var universalIdBuilder = TryWithImoNumber(imoNumber, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(imoNumber));
        return universalIdBuilder;
    }

    public UniversalIdBuilder TryWithImoNumber(in ImoNumber? imoNumber) => TryWithImoNumber(imoNumber, out _);

    public UniversalIdBuilder TryWithImoNumber(in ImoNumber? imoNumber, out bool succeeded)
    {
        if (imoNumber is null)
        {
            succeeded = false;
            return this;
        }
        succeeded = true;
        return this with { ImoNumber = imoNumber };
    }

    public UniversalIdBuilder WithoutImoNumber() => this with { ImoNumber = null };

    public sealed override int GetHashCode()
    {
        var hashCode = new HashCode();
        hashCode.Add(ImoNumber);
        hashCode.Add(LocalId);
        return hashCode.ToHashCode();
    }

    public override string ToString()
    {
        if (ImoNumber is null)
            throw new InvalidOperationException("Invalid Universal Id state: Missing IMO Number");
        if (LocalId is null)
            throw new InvalidOperationException("Invalid Universal Id state: Missing LocalId");

        string namingEntity = $"{NamingEntity}";
        using var lease = StringBuilderPool.Get();

        var builder = lease.Builder;

        builder.Append(namingEntity);
        builder.Append("/");
        builder.Append(ImoNumber.ToString());

        LocalId.ToString(builder);

        return lease.ToString();
    }
}
